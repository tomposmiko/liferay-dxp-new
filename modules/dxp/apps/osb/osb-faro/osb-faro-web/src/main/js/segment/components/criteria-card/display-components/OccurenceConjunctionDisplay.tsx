import React from 'react';
import {
	OCCURENCE_OPTIONS,
	RelationalOperators
} from 'segment/segment-editor/dynamic/utils/constants';

interface IOccurenceConjunctionDisplayProps {
	operatorName: RelationalOperators;
	value: number;
}

const OccurenceConjunctionDisplay: React.FC<IOccurenceConjunctionDisplayProps> = ({
	operatorName,
	value
}) => {
	const {label = ''} =
		OCCURENCE_OPTIONS.find(({value}) => value === operatorName) || {};

	return (
		<>
			<span>{label}</span>

			<b>{value}</b>

			<span>{Liferay.Language.get('times')}</span>
		</>
	);
};

export default OccurenceConjunctionDisplay;
