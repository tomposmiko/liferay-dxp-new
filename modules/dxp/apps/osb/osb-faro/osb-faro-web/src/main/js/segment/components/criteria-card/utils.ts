import moment from 'moment';
import {formatDateToTimeZone} from 'shared/util/date';
import {formatTime} from 'shared/util/time';
import {
	GEOLOCATION_OPTIONS,
	INPUT_DATE_FORMAT,
	INPUT_DISPLAY_DATE_TIME_FORMAT,
	isKnown,
	isUnknown,
	PropertyTypes,
	RelationalOperators,
	SUPPORTED_OPERATORS_MAP
} from 'segment/segment-editor/dynamic/utils/constants';

export function getOperatorLabel(
	operatorKey: string,
	type: PropertyTypes
): string {
	let supportedOperators;

	switch (type) {
		case PropertyTypes.AccountNumber:
		case PropertyTypes.AccountText:
			supportedOperators =
				SUPPORTED_OPERATORS_MAP[type.replace('account-', '')];
			break;
		case PropertyTypes.OrganizationBoolean:
		case PropertyTypes.OrganizationDate:
		case PropertyTypes.OrganizationDateTime:
		case PropertyTypes.OrganizationNumber:
		case PropertyTypes.OrganizationText:
			supportedOperators =
				SUPPORTED_OPERATORS_MAP[type.replace('organization-', '')];
			break;
		case PropertyTypes.SessionDateTime:
		case PropertyTypes.SessionNumber:
		case PropertyTypes.SessionText:
			supportedOperators =
				SUPPORTED_OPERATORS_MAP[type.replace('session-', '')];
			break;
		case PropertyTypes.SessionGeolocation:
			supportedOperators = GEOLOCATION_OPTIONS;
			break;
		case PropertyTypes.Behavior:
		case PropertyTypes.Boolean:
		case PropertyTypes.Date:
		case PropertyTypes.DateTime:
		case PropertyTypes.Duration:
		case PropertyTypes.Number:
		case PropertyTypes.OrganizationSelectText:
		case PropertyTypes.Text:
		default:
			supportedOperators = SUPPORTED_OPERATORS_MAP[type];
	}

	const operator = supportedOperators.find(
		({key, value}) => (key || value) === operatorKey
	);

	return operator ? operator.label : null;
}

export function maybeFormatToKnownType(
	operatorName: string,
	value: any
): string {
	const valueNull = value === null;

	if (operatorName === RelationalOperators.EQ && valueNull) {
		return isUnknown;
	} else if (operatorName === RelationalOperators.NE && valueNull) {
		return isKnown;
	}

	return operatorName;
}

export function maybeFormatValue(
	value: any,
	type: string,
	timeZoneId?: string
): string | number {
	switch (type) {
		case PropertyTypes.AccountText:
		case PropertyTypes.Behavior:
		case PropertyTypes.Interest:
		case PropertyTypes.SessionGeolocation:
		case PropertyTypes.SessionText:
		case PropertyTypes.Text:
			return `"${value}"`;
		case PropertyTypes.Boolean:
			return value.toUpperCase();
		case PropertyTypes.Date:
			return moment(value).format(INPUT_DATE_FORMAT);
		case PropertyTypes.DateTime:
		case PropertyTypes.SessionDateTime:
			return formatDateToTimeZone(
				value,
				INPUT_DISPLAY_DATE_TIME_FORMAT,
				timeZoneId
			);
		case PropertyTypes.Duration:
			return formatTime(value);
		case PropertyTypes.AccountNumber:
		case PropertyTypes.Number:
		case PropertyTypes.SessionNumber:
		default:
			return value;
	}
}
