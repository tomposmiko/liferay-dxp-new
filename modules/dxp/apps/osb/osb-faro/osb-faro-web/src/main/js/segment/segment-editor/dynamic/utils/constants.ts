import {TimeSpans} from 'shared/util/constants';

export const EVER = 'ever';
export const SINCE = 'since';

export const isKnown = 'is-known';
export const isUnknown = 'is-unknown';

/**
 * Constants for date formatting
 */

export const INPUT_DATE_FORMAT = 'YYYY-MM-DD';
export const INPUT_DATE_TIME_FORMAT = 'YYYY-MM-DDTHH:mmZ';
export const INPUT_DISPLAY_DATE_TIME_FORMAT = 'YYYY-MM-DD HH:mm';

/**
 * Constants for OData query.
 */

export enum Conjunctions {
	And = 'and',
	Or = 'or'
}

export enum CustomFunctionOperators {
	AccountsFilter = 'accounts-filter',
	AccountsFilterByCount = 'accounts-filter-by-count',
	ActivitiesFilter = 'activities-filter',
	ActivitiesFilterByCount = 'activities-filter-by-count',
	EventsFilterByCount = 'events-filter-by-count',
	InterestsFilter = 'interests-filter',
	OrganizationsFilter = 'organizations-filter',
	SessionsFilter = 'sessions-filter'
}

export enum DisplayOnlyOperators {
	IsKnown = 'ne',
	IsUnknown = 'eq'
}

export enum FunctionalOperators {
	Between = 'between',
	Contains = 'contains'
}

export enum NotOperators {
	NotAccountsFilter = 'not-accounts-filter',
	NotAccountsFilterByCount = 'not-accounts-filter-by-count',
	NotActivitiesFilter = 'not-activities-filter',
	NotActivitiesFilterByCount = 'not-activities-filter-by-count',
	NotContains = 'not-contains',
	NotEventsFilterByCount = 'not-events-filter-by-count',
	NotOrganizationsFilter = 'not-organizations-filter',
	NotSessionsFilter = 'not-sessions-filter'
}

export const GROUP = 'GROUP';

export enum RelationalOperators {
	EQ = 'eq',
	GE = 'ge',
	GT = 'gt',
	LE = 'le',
	LT = 'lt',
	NE = 'ne'
}

/**
 * Constants to match property types in the passed in supportedProperties array.
 */

export enum PropertyTypes {
	AccountNumber = 'account-number',
	AccountText = 'account-text',
	Behavior = 'behavior',
	Boolean = 'boolean',
	Date = 'date',
	DateTime = 'date-time',
	Duration = 'duration',
	Event = 'event',
	Interest = 'interest',
	Number = 'number',
	OrganizationBoolean = 'organization-boolean',
	OrganizationDate = 'organization-date',
	OrganizationDateTime = 'organization-date-time',
	OrganizationNumber = 'organization-number',
	OrganizationSelectText = 'organization-select-text',
	OrganizationText = 'organization-text',
	SelectText = 'select-text',
	SessionDateTime = 'session-date-time',
	SessionGeolocation = 'session-geolocation',
	SessionNumber = 'session-number',
	SessionText = 'session-text',
	Text = 'text'
}

/**
 * Constants for CriteriaBuilder component.
 */

export const CUSTOM_FUNCTION_OPERATOR_KEY_MAP = {
	['accounts.filter']: CustomFunctionOperators.AccountsFilter,
	['accounts.filterByCount']: CustomFunctionOperators.AccountsFilterByCount,
	['activities.filter']: CustomFunctionOperators.ActivitiesFilter,
	['activities.filterByCount']:
		CustomFunctionOperators.ActivitiesFilterByCount,
	['events.filterByCount']: CustomFunctionOperators.EventsFilterByCount,
	['interests.filter']: CustomFunctionOperators.InterestsFilter,
	['organizations.filter']: CustomFunctionOperators.OrganizationsFilter,
	['sessions.filter']: CustomFunctionOperators.SessionsFilter
};

export const SUPPORTED_CONJUNCTION_OPTIONS = [
	{
		key: Conjunctions.And,
		label: Liferay.Language.get('and'),
		name: Conjunctions.And
	},
	{
		key: Conjunctions.Or,
		label: Liferay.Language.get('or'),
		name: Conjunctions.Or
	}
];

export const SUPPORTED_OPERATORS_MAP = {
	[PropertyTypes.AccountNumber]: [
		{
			key: CustomFunctionOperators.AccountsFilter,
			label: Liferay.Language.get('is-fragment'),
			name: CustomFunctionOperators.AccountsFilter
		}
	],
	[PropertyTypes.AccountText]: [
		{
			key: CustomFunctionOperators.AccountsFilter,
			label: Liferay.Language.get('is-fragment'),
			name: CustomFunctionOperators.AccountsFilter
		}
	],
	[PropertyTypes.Behavior]: [
		{
			key: CustomFunctionOperators.ActivitiesFilterByCount,
			label: Liferay.Language.get('has-fragment'),
			name: CustomFunctionOperators.ActivitiesFilterByCount
		},
		{
			key: NotOperators.NotActivitiesFilterByCount,
			label: Liferay.Language.get('has-not-fragment'),
			name: NotOperators.NotActivitiesFilterByCount
		}
	],
	[PropertyTypes.Boolean]: [
		{
			key: RelationalOperators.EQ,
			label: Liferay.Language.get('is-fragment'),
			name: RelationalOperators.EQ
		}
	],
	[PropertyTypes.Date]: [
		{
			key: RelationalOperators.LT,
			label: Liferay.Language.get('is-before-fragment'),
			name: RelationalOperators.LT
		},
		{
			key: RelationalOperators.EQ,
			label: Liferay.Language.get('is-fragment'),
			name: RelationalOperators.EQ
		},
		{
			key: RelationalOperators.GT,
			label: Liferay.Language.get('is-after-fragment'),
			name: RelationalOperators.GT
		}
	],
	[PropertyTypes.DateTime]: [
		{
			key: RelationalOperators.LT,
			label: Liferay.Language.get('is-before-fragment'),
			name: RelationalOperators.LT
		},
		{
			key: RelationalOperators.EQ,
			label: Liferay.Language.get('is-fragment'),
			name: RelationalOperators.EQ
		},
		{
			key: RelationalOperators.GT,
			label: Liferay.Language.get('is-after-fragment'),
			name: RelationalOperators.GT
		}
	],
	[PropertyTypes.Duration]: [
		{
			key: RelationalOperators.GT,
			label: Liferay.Language.get('greater-than-fragment'),
			name: RelationalOperators.GT
		},
		{
			key: RelationalOperators.LT,
			label: Liferay.Language.get('less-than-fragment'),
			name: RelationalOperators.LT
		}
	],
	[PropertyTypes.Event]: [
		{
			key: CustomFunctionOperators.EventsFilterByCount,
			label: Liferay.Language.get('has-fragment'),
			name: CustomFunctionOperators.EventsFilterByCount
		},
		{
			key: NotOperators.NotActivitiesFilterByCount,
			label: Liferay.Language.get('has-not-fragment'),
			name: NotOperators.NotActivitiesFilterByCount
		}
	],
	[PropertyTypes.Interest]: [
		{
			key: CustomFunctionOperators.InterestsFilter,
			label: Liferay.Language.get('is-fragment'),
			name: CustomFunctionOperators.InterestsFilter
		}
	],
	[PropertyTypes.Number]: [
		{
			key: RelationalOperators.EQ,
			label: Liferay.Language.get('is-equal-to-fragment'),
			name: RelationalOperators.EQ
		},
		{
			key: RelationalOperators.GT,
			label: Liferay.Language.get('greater-than-fragment'),
			name: RelationalOperators.GT
		},
		{
			key: RelationalOperators.LT,
			label: Liferay.Language.get('less-than-fragment'),
			name: RelationalOperators.LT
		},
		{
			key: RelationalOperators.NE,
			label: Liferay.Language.get('is-not-equal-to-fragment'),
			name: RelationalOperators.NE
		},
		{
			key: isKnown,
			label: Liferay.Language.get('is-known-fragment'),
			name: DisplayOnlyOperators.IsKnown
		},
		{
			key: isUnknown,
			label: Liferay.Language.get('is-unknown-fragment'),
			name: DisplayOnlyOperators.IsUnknown
		}
	],
	[PropertyTypes.OrganizationBoolean]: [
		{
			key: CustomFunctionOperators.OrganizationsFilter,
			label: Liferay.Language.get('is-fragment'),
			name: CustomFunctionOperators.OrganizationsFilter
		}
	],
	[PropertyTypes.OrganizationDate]: [
		{
			key: CustomFunctionOperators.OrganizationsFilter,
			label: Liferay.Language.get('is-fragment'),
			name: CustomFunctionOperators.OrganizationsFilter
		}
	],
	[PropertyTypes.OrganizationDateTime]: [
		{
			key: CustomFunctionOperators.OrganizationsFilter,
			label: Liferay.Language.get('is-fragment'),
			name: CustomFunctionOperators.OrganizationsFilter
		}
	],
	[PropertyTypes.OrganizationNumber]: [
		{
			key: CustomFunctionOperators.OrganizationsFilter,
			label: Liferay.Language.get('is-fragment'),
			name: CustomFunctionOperators.OrganizationsFilter
		}
	],
	[PropertyTypes.OrganizationSelectText]: [
		{
			key: CustomFunctionOperators.OrganizationsFilter,
			label: Liferay.Language.get('is-fragment'),
			name: CustomFunctionOperators.OrganizationsFilter
		},
		{
			key: NotOperators.NotOrganizationsFilter,
			label: Liferay.Language.get('is-not-fragment'),
			name: NotOperators.NotOrganizationsFilter
		}
	],
	[PropertyTypes.OrganizationText]: [
		{
			key: CustomFunctionOperators.OrganizationsFilter,
			label: Liferay.Language.get('is-fragment'),
			name: CustomFunctionOperators.OrganizationsFilter
		}
	],
	[PropertyTypes.SelectText]: [
		{
			key: RelationalOperators.EQ,
			label: Liferay.Language.get('is-fragment'),
			name: RelationalOperators.EQ
		},
		{
			key: RelationalOperators.NE,
			label: Liferay.Language.get('is-not-fragment'),
			name: RelationalOperators.NE
		}
	],
	[PropertyTypes.SessionDateTime]: [
		{
			key: CustomFunctionOperators.SessionsFilter,
			label: Liferay.Language.get('is-fragment'),
			name: CustomFunctionOperators.SessionsFilter
		}
	],
	[PropertyTypes.SessionGeolocation]: [
		{
			key: CustomFunctionOperators.SessionsFilter,
			label: Liferay.Language.get('is-fragment'),
			name: CustomFunctionOperators.SessionsFilter
		}
	],
	[PropertyTypes.SessionNumber]: [
		{
			key: CustomFunctionOperators.SessionsFilter,
			label: Liferay.Language.get('is-fragment'),
			name: CustomFunctionOperators.SessionsFilter
		}
	],
	[PropertyTypes.SessionText]: [
		{
			key: CustomFunctionOperators.SessionsFilter,
			label: Liferay.Language.get('is-fragment'),
			name: CustomFunctionOperators.SessionsFilter
		}
	],
	[PropertyTypes.Text]: [
		{
			key: RelationalOperators.EQ,
			label: Liferay.Language.get('is-fragment'),
			name: RelationalOperators.EQ
		},
		{
			key: RelationalOperators.NE,
			label: Liferay.Language.get('is-not-fragment'),
			name: RelationalOperators.NE
		},
		{
			key: FunctionalOperators.Contains,
			label: Liferay.Language.get('contains-fragment'),
			name: FunctionalOperators.Contains
		},
		{
			key: NotOperators.NotContains,
			label: Liferay.Language.get('does-not-contain-fragment'),
			name: NotOperators.NotContains
		},
		{
			key: isKnown,
			label: Liferay.Language.get('is-known-fragment'),
			name: DisplayOnlyOperators.IsKnown
		},
		{
			key: isUnknown,
			label: Liferay.Language.get('is-unknown-fragment'),
			name: DisplayOnlyOperators.IsUnknown
		}
	]
};

export const SUPPORTED_PROPERTY_TYPES_MAP = {
	[PropertyTypes.AccountNumber]: [CustomFunctionOperators.AccountsFilter],
	[PropertyTypes.AccountText]: [CustomFunctionOperators.AccountsFilter],
	[PropertyTypes.Behavior]: [
		CustomFunctionOperators.ActivitiesFilterByCount,
		NotOperators.NotActivitiesFilterByCount
	],
	[PropertyTypes.Boolean]: [RelationalOperators.EQ],
	[PropertyTypes.Date]: [
		RelationalOperators.EQ,
		RelationalOperators.GE,
		RelationalOperators.GT,
		RelationalOperators.LE,
		RelationalOperators.LT,
		RelationalOperators.NE
	],
	[PropertyTypes.DateTime]: [
		RelationalOperators.EQ,
		RelationalOperators.GE,
		RelationalOperators.GT,
		RelationalOperators.LE,
		RelationalOperators.LT,
		RelationalOperators.NE
	],
	[PropertyTypes.Duration]: [RelationalOperators.GT, RelationalOperators.LT],
	[PropertyTypes.Event]: [
		CustomFunctionOperators.EventsFilterByCount,
		NotOperators.NotActivitiesFilterByCount
	],
	[PropertyTypes.Interest]: [CustomFunctionOperators.InterestsFilter],
	[PropertyTypes.Number]: [
		RelationalOperators.EQ,
		RelationalOperators.GE,
		RelationalOperators.GT,
		RelationalOperators.LE,
		RelationalOperators.LT,
		RelationalOperators.NE
	],
	[PropertyTypes.SessionDateTime]: [PropertyTypes.SessionDateTime],
	[PropertyTypes.SessionNumber]: [CustomFunctionOperators.SessionsFilter],
	[PropertyTypes.SessionText]: [CustomFunctionOperators.SessionsFilter],
	[PropertyTypes.Text]: [
		RelationalOperators.EQ,
		RelationalOperators.NE,
		FunctionalOperators.Contains,
		NotOperators.NotContains,
		DisplayOnlyOperators.IsKnown,
		DisplayOnlyOperators.IsUnknown
	]
};

/**
 * Values for criteria row inputs.
 */

export const STRING_OPTIONS = [
	FunctionalOperators.Contains,
	NotOperators.NotContains,
	RelationalOperators.EQ,
	RelationalOperators.NE
];

export const STRING_OPERATOR_LABELS_MAP = {
	[FunctionalOperators.Contains]: Liferay.Language.get('contains-fragment'),
	[NotOperators.NotContains]: Liferay.Language.get('not-contains-fragment'),
	[RelationalOperators.EQ]: Liferay.Language.get('is-fragment'),
	[RelationalOperators.NE]: Liferay.Language.get('is-not-fragment')
};

export const BOOLEAN_OPTIONS = [
	{
		label: Liferay.Language.get('true'),
		value: 'true'
	},
	{
		label: Liferay.Language.get('false'),
		value: 'false'
	}
];

export const INTEREST_BOOLEAN_OPTIONS = [
	{
		label: Liferay.Language.get('is-fragment'),
		value: 'true'
	},
	{
		label: Liferay.Language.get('is-not-fragment'),
		value: 'false'
	}
];

export const OCCURENCE_OPTIONS = [
	{
		key: RelationalOperators.GE,
		label: Liferay.Language.get('at-least-fragment'),
		value: RelationalOperators.GE
	},
	{
		key: RelationalOperators.LE,
		label: Liferay.Language.get('at-most-fragment'),
		value: RelationalOperators.LE
	}
];

export const GEOLOCATION_OPTIONS = [
	{
		label: Liferay.Language.get('was-fragment'),
		value: RelationalOperators.EQ
	},
	{
		label: Liferay.Language.get('was-not-fragment'),
		value: RelationalOperators.NE
	},
	{
		label: Liferay.Language.get('contained-fragment'),
		value: FunctionalOperators.Contains
	},
	{
		label: Liferay.Language.get('did-not-contain-fragment'),
		value: NotOperators.NotContains
	}
];

export const TIME_CONJUNCTION_OPTIONS = [
	{
		label: Liferay.Language.get('since-fragment'),
		value: SINCE
	},
	{
		label: Liferay.Language.get('after-fragment'),
		value: RelationalOperators.GT
	},
	{
		label: Liferay.Language.get('before-fragment'),
		value: RelationalOperators.LT
	},
	{
		label: Liferay.Language.get('between-fragment'),
		value: FunctionalOperators.Between
	},
	{
		label: Liferay.Language.get('ever-fragment'),
		value: EVER
	},
	{
		label: Liferay.Language.get('on-fragment'),
		value: RelationalOperators.EQ
	}
];

export const ACTIVITY_KEY = 'activityKey';
export const EVENT_KEY = 'eventDefinitionId';

export const TIME_PERIOD_OPTIONS = [
	{
		label: Liferay.Language.get('last-24-hours'),
		value: TimeSpans.Last24Hours
	},
	{
		label: Liferay.Language.get('yesterday'),
		value: TimeSpans.Yesterday
	},
	{
		label: Liferay.Language.get('last-seven-days'),
		value: TimeSpans.Last7Days
	},
	{
		label: Liferay.Language.get('last-28-days'),
		value: TimeSpans.Last28Days
	},
	{
		label: Liferay.Language.get('last-30-days'),
		value: TimeSpans.Last30Days
	},
	{
		label: Liferay.Language.get('last-90-days'),
		value: TimeSpans.Last90Days
	}
];

export {TimeSpans};
