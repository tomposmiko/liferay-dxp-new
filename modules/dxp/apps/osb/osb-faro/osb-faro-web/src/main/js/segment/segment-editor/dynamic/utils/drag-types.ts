export enum DragTypes {
	CriteriaGroup = 'criteria-group',
	CriteriaRow = 'criteria-row',
	Property = 'property'
}
