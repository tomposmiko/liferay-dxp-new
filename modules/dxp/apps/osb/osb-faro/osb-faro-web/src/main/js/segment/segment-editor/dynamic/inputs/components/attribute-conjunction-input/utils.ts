import {BetweenNumber} from '../BetweenNumberInput';
import {
	BOOLEAN_LABELS_MAP,
	DATE_OPERATOR_LONGHAND_LABELS_MAP,
	DATE_OPTIONS,
	DURATION_OPERATOR_LONGHAND_LABELS_MAP,
	DURATION_OPTIONS,
	NUMBER_OPERATOR_LONGHAND_LABELS_MAP,
	NUMBER_OPTIONS
} from 'event-analysis/utils/utils';
import {DataTypes} from 'event-analysis/utils/types';
import {DateRange} from 'shared/components/DateRangeInput';
import {
	FunctionalOperators,
	RelationalOperators,
	STRING_OPERATOR_LABELS_MAP,
	STRING_OPTIONS
} from '../../../utils/constants';
import {isNumber} from 'lodash';
import {isValid} from '../../../utils/utils';

export const createOption = (option, dataType: DataTypes) => {
	const LABELS_MAP = {
		[DataTypes.Boolean]: BOOLEAN_LABELS_MAP,
		[DataTypes.Date]: DATE_OPERATOR_LONGHAND_LABELS_MAP,
		[DataTypes.Duration]: DURATION_OPERATOR_LONGHAND_LABELS_MAP,
		[DataTypes.Number]: NUMBER_OPERATOR_LONGHAND_LABELS_MAP,
		[DataTypes.String]: STRING_OPERATOR_LABELS_MAP // STRING_OPERATOR_LABELS_MAP is provided from the segment-editor utils as "NotContains" differs from segment-editor and event-analysis. We should be able to use the evente-analysis version once we move away from odata.
	};

	return {
		label: LABELS_MAP[dataType][option],
		value: option
	};
};

export const getOperatorOptions = (dataType: DataTypes) => {
	const OPERATOR_OPTIONS = {
		[DataTypes.Date]: DATE_OPTIONS,
		[DataTypes.Duration]: DURATION_OPTIONS,
		[DataTypes.Number]: NUMBER_OPTIONS,
		[DataTypes.String]: STRING_OPTIONS // STRING_OPTIONS is provided from the segment-editor utils as "NotContains" differs from segment-editor and event-analysis. We should be able to use the evente-analysis version once we move away from odata.
	};

	return OPERATOR_OPTIONS[dataType]?.map(option =>
		createOption(option, dataType)
	);
};

export const getDefaultAttributeOperator = (
	dataType: DataTypes
): RelationalOperators | FunctionalOperators => {
	switch (dataType) {
		case DataTypes.Boolean:
		case DataTypes.Date:
			return RelationalOperators.EQ;
		case DataTypes.Duration:
		case DataTypes.Number:
			return RelationalOperators.GT;
		case DataTypes.String:
		default:
			return FunctionalOperators.Contains;
	}
};

export const getDefaultAttributeValue = (
	dataType: DataTypes,
	operatorName: RelationalOperators | FunctionalOperators
): string | {end: number | string; start: number | string} => {
	if (
		operatorName === FunctionalOperators.Between &&
		[DataTypes.Number, DataTypes.Date].filter(type => type === dataType)
	) {
		return {end: '', start: ''};
	}

	switch (dataType) {
		case DataTypes.Boolean:
			return 'true';
		case DataTypes.Date:
		case DataTypes.Number:
		case DataTypes.Duration:
		case DataTypes.String:
		default:
			return '';
	}
};

export const validateAttributeValue = (
	value: string | number | BetweenNumber | DateRange,
	dataType: DataTypes,
	operatorName?: FunctionalOperators | RelationalOperators
): boolean => {
	if (
		operatorName === FunctionalOperators.Between &&
		[DataTypes.Number, DataTypes.Date].filter(type => type === dataType)
	) {
		const {end, start} = value as BetweenNumber;

		return isValid(end) && isValid(start);
	}

	switch (dataType) {
		case DataTypes.Boolean:
			return value === 'true' || value === 'false';
		case DataTypes.Duration:
		case DataTypes.Number:
			return isNumber(value);
		case DataTypes.Date:
		case DataTypes.String:
		default:
			return isValid(value);
	}
};
