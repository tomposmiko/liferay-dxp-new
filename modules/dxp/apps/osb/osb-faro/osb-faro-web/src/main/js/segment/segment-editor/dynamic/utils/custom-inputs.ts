import {createNewGroup} from './utils';
import {CustomValue} from 'shared/util/records';
import {fromJS, Map} from 'immutable';
import {isArray} from 'lodash';
import {TIME_PERIOD_OPTIONS} from './constants';

/**
 * Create the valueIMap for a custom input.
 * @param {Array} params - Array of root keys for the valueIMap.
 * @returns {CustomValue} The params array converted into a deeply immutable valueIMap.
 */
export const createCustomValueMap = (
	params: {key: string; value: any}[]
): CustomValue => {
	let valueIMap = Map();

	params.forEach(({key, value}) => {
		if (isArray(value)) {
			valueIMap = valueIMap.set(key, fromJS(createNewGroup(value)));
		} else {
			valueIMap = valueIMap.set(key, value);
		}
	});

	return new CustomValue(valueIMap);
};

/**
 * Get the filter Criterion Map at index.
 * @param {CustomValue} valueIMap - The Immutable Map representing the custom input value.
 * @param {number} index - The index of the Criterion Map.
 * @returns {Map} The Criterion Map.
 */
export const getFilterCriterionIMap = (valueIMap: CustomValue, index: number) =>
	valueIMap.getIn(['criterionGroup', 'items', index]);

/**
 * Get the index where the first matching property name is found in the criteria list.
 * This is useful if we don't know where the criterion is at in the list, but we know the property name.
 * @param {CustomValue} valueIMap - The Immutable Map representing the custom input value.
 * @returns {number} The index of the matching criterion or -1 if not found.
 */
export const getIndexFromPropertyName = (
	valueIMap: CustomValue,
	propertyName: string
): number =>
	valueIMap
		.getIn(['criterionGroup', 'items'])
		.findIndex(entry => entry.get('propertyName') === propertyName);

/**
 * Get the operator name from the criterion at the specified index in the valueIMap.
 * @param {CustomValue} valueIMap - The Immutable Map representing the custom input value.
 * @returns {string} The operator name of the criterion at the specified index.
 */
export const getOperator = (valueIMap: CustomValue, index: number): string =>
	getPropertyValue(valueIMap, 'operatorName', index);

/**
 * Get the time period value from the valueIMap.
 * @param {CustomValue} valueIMap - The Immutable Map representing the custom input value.
 * @returns {string} The time period value from the criterion with a 'completeDate' property name.
 */
export const getCompleteDate = (valueIMap: CustomValue): string => {
	const index = getIndexFromPropertyName(valueIMap, 'completeDate');

	return getPropertyValue(valueIMap, 'value', index);
};

/**
 * Get the value of the propertyName at the specified index.
 * @param {CustomValue} valueIMap - The Immutable Map representing the custom input value.
 * @param {string} propertyName - The propertyName string in the criterion to update.
 * @param {number} index - The index of the criterion in the items list.
 * @returns {*} The value of the criterion propertyName at the specified index in the items list.
 */
export const getPropertyValue = (
	valueIMap: CustomValue,
	propertyName: string,
	index: number
): any => valueIMap.getIn(['criterionGroup', 'items', index, propertyName]);

/**
 * Get the time period label from the time period value.
 * @param {string} timePeriod - The time period value.
 * @returns {string} The time period label that matches the timePeriodValue.
 */
export const getTimePeriodLabel = (value: string): string => {
	const timePeriod = TIME_PERIOD_OPTIONS.find(
		timePeriod => timePeriod.value === value
	);

	return timePeriod ? timePeriod.label : null;
};

/**
 * Remove entries in valueIMap by their index in the criteria list.
 * @param {CustomValue} valueIMap - The Immutable Map representing the custom input value.
 * @param {Array} indexArray - Array of indexes to remove from the criteria list in valueIMap.
 * @returns {CustomValue} The valueIMap with the items at the indexes in indexArray removed.
 */
export const removeItemsByIndex = (
	valueIMap: CustomValue,
	indexArray: number[]
): CustomValue =>
	valueIMap.updateIn(['criterionGroup', 'items'], iList =>
		iList.filterNot((_, i) => indexArray.includes(i))
	) as CustomValue;

/**
 * Set the operator name from the criterion at the specified index in the valueIMap.
 * @param {CustomValue} valueIMap - The Immutable Map representing the custom input value.
 * @param {number} index - The index of the criterion in the items list.
 * @param {*} value - The value to update in the valueIMap.
 * @returns {CustomValue} The updated valueIMap.
 */
export const setOperator = (
	valueIMap: CustomValue,
	index: number,
	value: any
): CustomValue => setPropertyValue(valueIMap, 'operatorName', index, value);

/**
 * Set the time period value in the valueIMap.
 * @param {CustomValue} valueIMap - The Immutable Map representing the custom input value.
 * @param {string} value - The value to update for the time period.
 * @returns {CustomValue} The updated valueIMap.
 */
export const setCompleteDate = (
	valueIMap: CustomValue,
	completeDate: string
): CustomValue => {
	const index = getIndexFromPropertyName(valueIMap, 'completeDate');

	return setPropertyValue(valueIMap, 'value', index, completeDate);
};

/**
 * Set the value of the propertyName at the specified index.
 * @param {CustomValue} valueIMap - The Immutable Map representing the custom input value.
 * @param {string} propertyName - The propertyName string in the criterion to update.
 * @param {number} index - The index of the criterion in the items list.
 * @param {*} value - The value to update in the valueIMap.
 * @returns {CustomValue} The updated valueIMap.
 */
export const setPropertyValue = (
	valueIMap: CustomValue,
	propertyName: string,
	index: number,
	value: any
): CustomValue =>
	valueIMap.setIn(
		['criterionGroup', 'items', index, propertyName],
		value
	) as CustomValue;
