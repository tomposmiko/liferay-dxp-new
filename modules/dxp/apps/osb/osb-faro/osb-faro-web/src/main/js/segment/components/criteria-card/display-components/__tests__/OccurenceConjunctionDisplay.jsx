import OccurenceConjunctionDisplay from '../OccurenceConjunctionDisplay';
import React from 'react';
import {RelationalOperators} from 'segment/segment-editor/dynamic/utils/constants';
import {render} from '@testing-library/react';

jest.unmock('react-dom');

describe('OccurenceConjunctionDisplay', () => {
	it('should render', () => {
		const {container} = render(
			<OccurenceConjunctionDisplay
				operatorName={RelationalOperators.GT}
				value={13}
			/>
		);

		expect(container).toMatchSnapshot();
	});
});
