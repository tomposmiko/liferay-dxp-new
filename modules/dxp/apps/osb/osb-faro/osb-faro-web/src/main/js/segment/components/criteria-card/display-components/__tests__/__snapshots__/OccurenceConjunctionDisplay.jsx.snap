// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OccurenceConjunctionDisplay should render 1`] = `
<div>
  <span />
  <b>
    13
  </b>
  <span>
    time(s)
  </span>
</div>
`;
